torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -373,92 +364,317 @@ class _ExecOrderWarnStatus(Enum):
-            _get_param_to_unflat_param_names(root_module)
+            _get_param_to_unflat_param_names(fsdp_root),
         )
+        # TODO (awgu): We can broadcast the metadata of rank 0's `all_handles`
+        # to check that all ranks have the same handles in the same order.
+        # https://github.com/pytorch/pytorch/issues/79620
 
-    def get_param_index(self, param: FlatParameter) -> int:
-        """Returns a unique non-negative parameter index for ``param`` if it is
-        valid or -1 otherwise. Critically, this index assignment must be the
-        same across ranks."""
-        assert isinstance(param, FlatParameter), \
-            f"Expects `param` is a `FlatParameter` but got {type(param)}"
-        for i, p in enumerate(self._all_flat_params):
-            if p is param:
-                return i
-        return -1
-
-    def get_param(self, param_index: int) -> Optional[FlatParameter]:
-        """Returns the parameter corresponding to ``param_index`` or ``None``
-        if the index is invalid."""
-        for i, p in enumerate(self._all_flat_params):
-            if i == param_index:
-                return p
-        return None
+    def get_handles_to_backward_prefetch(
+        self,
+        current_handles_key: Tuple[FlatParamHandle, ...],
+    ) -> Optional[Tuple[FlatParamHandle, ...]]:
+        """
+        Returns the handles key of the handles to backward prefetch given the
+        current handles key or ``None`` if there is no valid handles key to
+        prefetch.
+        """
+        current_index = self.handles_to_post_forward_order_index.get(current_handles_key, None)
+        if current_index is None:
+            return None
+        target_index = current_index - 1
+        if target_index < 0:
+            return None
+        target_handles_key = self.handles_post_forward_order[target_index]
+        return target_handles_key
 
-    def get_unflat_param_names(self, param_index: int) -> List[str]:
-        """Returns a :class:`list` of unflattened parameter names comprising
-        the flattened parameter with index ``param_index`` or an empty
-        :class:`list` if ``param_index`` is invalid."""
-        param = self.get_param(param_index)
-        if param is None:
-            return []
-        assert param in self._param_to_unflat_param_names, \
-            "Internal data structures out of sync; check `init()`"
-        return self._param_to_unflat_param_names[param]
+    def record_post_forward(self, handles: List[FlatParamHandle]) -> None:
+        """
+        Records ``handles`` in the post-forward order, where ``handles`` should
+        be a group of handles used in the same module's forward.
+
+        Unlike :meth:`record_pre_forward`, this records the order *every*
+        iteration with the expectation that the recorded order is reset in
+        :meth:`next_iter`, and the recorded order includes empty handles keys.
+        """
+        handles_key = tuple(handles)
```

Comment:
TODO: After offline discussion with @zhaojuanmao, we will omit any "bubbles" (i.e. empty handles keys).
```suggestion
        if not handles:
            return
        handles_key = tuple(handles)
```
I will leave this comment here until I push a change.